### PR TITLE
[Windows support] Add Clang downloads for Windows

### PR DESCRIPTION
**Pull request by micbou**
`7-zip` is the best way I found to obtain the files from the Clang installer without installing it.

### CODE REVIEW

File: cpp/ycm/CMakeLists.txt [hunk 1]
```diff
@@ -103,16 +114,25 @@ if ( USE_CLANG_COMPLETER AND
     execute_process( COMMAND tar -xjf ${CLANG_FILENAME} )
   elseif( CLANG_FILENAME MATCHES ".+xz" )
     execute_process( COMMAND tar -xJf ${CLANG_FILENAME} )
+  elseif( CLANG_FILENAME MATCHES ".+exe" )
+    # NSIS tool is used to generate the Clang installer on Windows. Files
+    # from the installer can be extracted with 7-Zip.
+    execute_process( COMMAND 7z x ${CLANG_FILENAME} OUTPUT_QUIET )
```
Comment (puremourning) [inline cpp/ycm/CMakeLists.txt]: Does this make 7-zip a requirement for installation on Windows? If so do we need to document somewhere that 7-zip is a requirement for installing YCM?

Comment (micbou) [inline cpp/ycm/CMakeLists.txt]: Yes, it is a requirement for building YCM with Clang support. I will document it in the `Windows Installation` section of the YCM `README.md`.

Comment (Valloric) [inline cpp/ycm/CMakeLists.txt]: Any chance we could not have that extra dependency? I could live with another python-only dependency in `third_party` that could do the necessary work; but we'd be shipping that then and the user wouldn't have to install it themselves.

Comment (Valloric) [inline cpp/ycm/CMakeLists.txt]: In other words, it's the user-level dependency of "go install this other thing first" that I'd like to avoid.

**Comment by micbou** [comment]
I removed the `7-zip` dependency. New approach is to (silently) execute the Clang installer, copy the installed files in a new directory, and uninstall it to clean up the registry. There are minor drawbacks:
 - when Clang is already installed, a dialog window will appear asking the user if he want to uninstall the previous installation;
 - 2 console windows appear a very short time during the build;
 - appveyor runs takes 1-2 more minutes.

**Comment by Valloric** [comment]
@micbou I appreciate you trying to remove the 7zip depedency, but I think the new approach with running the installer is a step backwards from that. The issues with dialogs and the console windows plus the somewhat hacky approach of the whole thing seem... unfortunate.

Frankly, I'd rather have the 7zip dep if nothing better is available.

**Comment by Valloric** [comment]
You know what? Let's just go with the installers.

**Comment by vheon** [comment]
I'd rather not play with the user installed packages. I would prefer the 7-zip dep.

**Comment by Valloric** [comment]
@vheon I agree with you, but we also have two suboptimal approach to chose from. At this point, I'm fine with whichever ends up being less hassle for us. I'll let @micbou decide since he'll be maintaining the windows support. Whatever works best for him, we'll go with that.

**Comment by micbou** [comment]
Clearly, the 7-zip approach is cleaner. 7-zip is also one of the most popular file archiver on Windows so I don't think this is a problem to require it. What I could improve is to avoid the need to add it to the PATH. I will work on this tomorrow.

**Comment by micbou** [comment]
I reverted the PR to the 7-zip approach and improved it. The 7-zip executable is now automatically found searching the PATH and using the registry key. An error is raised if no executable is found.

**Comment by vheon** [comment]
I like this version much better.

**Comment by Valloric** [comment]
Nice, thanks!

@homu r+

**Comment by micbou** [comment]
@homu retry

**Comment by micbou** [comment]
@homu r+

**Pull request merged**